pi4j-core/src/main/java/com/pi4j/io/gpio/digital/DigitalOutputBase.java

```diff
@@ -128,10 +128,100 @@ public Future<?> pulseAsync(int interval, TimeUnit unit, DigitalState state, Cal
+     *
+     * Representation:
+     *
+     *   HIGH +-----+     +-----+     +-----+
+     *        |     |     |     |     |     |
+     *   LOW  +     +-----+     +-----+     +-----+
+     *        ^                                   ^
+     * start -┘                                   └- stop
+     *         \___/ \___/
+     *         delay  delay
+     *
+     *        \___________________________________/
+     *                      duration
+     *
+     * Example: Delay = 1 sec / duration = 10
+     *          Output will be like so (suppose the initial state is ON):
+     *          1 - 0 - 1 - 0 - 1 - 0 - 1 - 0 - 1 - 0 with each state lasting for 1 second.
+     *          So, if you would connect a LED to the pin, you would see the LED switching
+     *          on and off for 5 times, NOT 10 times!!!
+     *
+     * @param unit - The time unit used to calculate the delay.
+     * @param state - The initial state of the pin.
+     * @param callback - The method to call, if any, once the blinking is done.
+     * @return - The DigitalOutput object itself.
+     */
     @Override
     public DigitalOutput blink(int delay, int duration, TimeUnit unit, DigitalState state, Callable<Void> callback) {
-        // TODO :: IMPLEMENT DIGITAL OUTPUT BLINK
-        throw new UnsupportedOperationException("BLINK has not yet been implemented!");
+        int millis = 0;
+
+        if (delay <= 0) {
+            throw new IllegalArgumentException("A delay of zero or less is not supported.");
+        }
+        if (duration <= 0) {
+            throw new IllegalArgumentException("A duration of zero or less is not supported.");
+        }
+
+        switch (unit) {
+            case MICROSECONDS:
+                throw new IllegalArgumentException("TimeUnit.MICROSECONDS is not supported.");
+
+            case MILLISECONDS:
+                millis = delay;
+                break;
+
+            case SECONDS:
+                millis = (delay * 1000);
+                break;
+
+            case MINUTES:
+                millis = (delay * 1000 * 60);
+                break;
+
+            case HOURS:
+                millis = (delay * 1000 * 60 * 60);
+                break;
+
+            default:
+                throw new IllegalArgumentException("TimeUnit provided is not supported.");
+        }
+
+        this.state(state);
+
+        for (int i = 0; i < duration; i++) {
+            // block the current thread for the pulse duration
+            try {
+                Thread.sleep(millis);
+            }
+            catch (InterruptedException e) {
+                throw new RuntimeException("Pulse blocking thread interrupted. Exception message: [" + e.getMessage() + "].");
+            }
+
+
+            // toggle the pulse state
+            this.state(DigitalState.getInverseState(this.state));
+        }
+
+        // invoke callback if one was defined
+        if (callback != null) {
```

Comment:
If this method is blocking, then this callback isn't really useful, but if it would be non-blocking, then this would be excellent.

Reply:
Well, maybe, maybe not.  In any case the caller will be informed about the end of the blocking phase and can still do whatever (s)he wants in the callback method.